### PR TITLE
CodeMirror blob: Wrap `stencil` tokens in links to definitions

### PR DESCRIPTION
**Pull request by umpox**
https://user-images.githubusercontent.com/9516420/195355312-355d038c-8d75-4aff-a347-dda1ffab9bc8.mp4

## Description

This is a follow up to https://github.com/sourcegraph/sourcegraph/pull/42936 that now enables keyboard controls for directly navigation to a definition.

**Prefetching definitions**

This is quite a shift from our current setup, right now we only fetch definition/references/hover information when a possible token is _hovered_. This PR changes that so, when we have `stencil` ranges, we prefetch all definition information within the current viewport. We don't have a backend API that supports that yet (created https://github.com/sourcegraph/sourcegraph/issues/43179 to address this), but this PR batches the definitions into a single GQL query so we don't suffer too much on client perf. I've gated this behind an experimental feature, although I wouldn't object to removing this if there's no backend concerns.

## Test plan

Tested locally

`sg start web-standalone`

Navigate to a file with precise code intel like: https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/backend/external_services.go?L31

### CODE REVIEW

File: client/web/src/repo/blob/BlobPage.tsx [hunk 1]
```diff
@@ -107,9 +107,10 @@ export const BlobPage: React.FunctionComponent<React.PropsWithChildren<BlobPageP
     const enableLazyBlobSyntaxHighlighting = useExperimentalFeatures(
         features => features.enableLazyBlobSyntaxHighlighting ?? false
     )
-    const enableTokenKeyboardNavigation =
+    let enableTokenKeyboardNavigation =
```
Comment (umpox) [inline client/web/src/repo/blob/BlobPage.tsx:110]: Temporarily enabled for this PR for easy testing with `web-standalone`

File: client/web/src/repo/blob/CodeMirrorBlob.tsx [hunk 2]
```diff
@@ -180,7 +171,13 @@ export const Blob: React.FunctionComponent<BlobProps> = props => {
                 initialSelection: position.line !== undefined ? position : null,
                 navigateToLineOnAnyClick: navigateToLineOnAnyClick ?? false,
             }),
-            tokenKeyboardNavigation ? tokensAsLinks.of({ history, links: tokenLinks }) : [],
+            tokenKeyboardNavigation
+                ? tokensAsLinks({
+                      history,
+                      blobInfo,
+                      preloadGoToDefinition: true,
```
Comment (umpox) [inline client/web/src/repo/blob/CodeMirrorBlob.tsx:178]: Temporarily enabled for this PR for easy testing with web-standalone

File: client/web/src/repo/blob/definitions.ts [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            blob: {
+                lsif: {
+                    [key: string]: {
+                        nodes: DefinitionFields[]
+                    }
+                }
+            }
+        }
+    }
+}
```
Comment (umpox) [inline client/web/src/repo/blob/definitions.ts:52]: As far as I can tell, there's no way to make GQL codegen happy when building a query like this that uses GQL aliases. The best solution I came to was to use a fragment within each alias, and then manually type the rest of the query like we do here

File: client/web/src/repo/blob/BlobPage.tsx [hunk 4]
```diff
@@ -107,9 +107,10 @@ export const BlobPage: React.FunctionComponent<React.PropsWithChildren<BlobPageP
...
         isSettingsValid(props.settingsCascade) &&
         props.settingsCascade.final['codeIntel.blobKeyboardNavigation'] === 'token'
+    enableTokenKeyboardNavigation = true
```
Comment (olafurpg) [inline client/web/src/repo/blob/BlobPage.tsx:113]: TODO: remove before merging

**Review by olafurpg — APPROVED** [review]
This is incredibly game changing, I love using the product with cursor navigation 😍 A few raw observations after trying this out:


- Clicking on the definition itself should open "find references". Not super important, but if there's only one reference, it would redirect to the single reference.
- When the definition is already visible in the viewport, there's no visual hint when you click on the definition twice. It would be nice if the selected line wiggles or something like that to draw your attention to it.
- I'd prefer to use up/down/left/right arrow keys over tab/shift-tab. One surprising behavior with tab/shift-tab is that the focus on elements outside of the blob (esp. easy with a fast key repeat rate and holding down tab/shift-tab). 
- The token cursor position should ideally be preserved in session storage per blob so that I can navigate to the definition with Enter, and use Cmd+Left to quickly go back to the same position I was in earlier.
- It would be nice to be able to load the popover with `Space` or something like that
- This feature desperately needs to be able to work with search-based intel to have a huge impact, but it's a fine starting point IMO to only support precise intel. I think it's OK to ship this as GA with only precise intel support, but there's a chance not everybody agrees with that. This is a prime example of the cool use-cases we can support with precise intel that's more fuzzy with search-based.

**Pull request merged**